node/mock-test.ts

```diff
@@ -123,4 +128,186 @@ export class MockTestRunner {
+                break;
+            case 6:
+                downloadVersion = 'v6.10.3';
+                break;
+            case 10:
+                downloadVersion = 'v10.15.1';
+                break;
+            default:
+                throw new Error('Invalid node version, must be 5, 6, or 10 (received ' + version + ')');
+        }
+
+        // Install node in home directory if it isn't already there.
+        const downloadDestination: string = path.join(downloadDirectory, 'node' + version);
+        const pathToExe: string = this.getPathToNodeExe(downloadVersion, downloadDestination);
+        if (pathToExe) {
+            return pathToExe;
+        }
+        else {
+            return this.downloadNode(downloadVersion, downloadDestination);
+        }
+    }
+
+    // Determines the correct version of node to use based on the contents of the task's task.json. Defaults to Node 10.
+    private getNodeVersion(): number {
+        const taskJsonPath: string = this.getTaskJsonPath();
+        if (!taskJsonPath) {
+            console.warn('Unable to find task.json, defaulting to use Node 10');
+            return 10;
+        }
+        const taskJsonContents = fs.readFileSync(taskJsonPath, { encoding: 'utf-8' });
+        const taskJson: object = JSON.parse(taskJsonContents);
+
+        let nodeVersionFound = false;
+        const execution: object = taskJson['execution'];
+        const keys = Object.keys(execution);
+        for (let i = 0; i < keys.length; i++) {
+            if (keys[i].toLowerCase() == 'node10') {
+                // Prefer node 10 and return immediately.
+                return 10;
+            }
+            else if (keys[i].toLowerCase() == 'node') {
+                nodeVersionFound = true;
+            }
+        }
+
+        if (!nodeVersionFound) {
+            console.warn('Unable to determine execution type from task.json, defaulting to use Node 10');
+            return 10;
+        }
+
+        return 6;
+    }
+
+    // Returns the path to the task.json for the task being tested. Returns null if unable to find it.
+    // Searches by moving up the directory structure from the initial starting point and checking at each level.
+    private getTaskJsonPath(): string {
+        if (this._taskJsonPath) {
+            return this._taskJsonPath;
+        }
+        let curPath: string = this._testPath;
+        let newPath: string = path.join(this._testPath, '..');
+        while (curPath != newPath) {
+            curPath = newPath;
+            let taskJsonPath: string = path.join(curPath, 'task.json');
+            if (fs.existsSync(taskJsonPath)) {
+                return taskJsonPath;
+            }
+            newPath = path.join(curPath, '..');
+        }
+        return '';
+    }
+
+    // Downloads the specified node version to the download destination. Returns a path to node.exe
+    private downloadNode(nodeVersion: string, downloadDestination: string): string {
```

Comment:
should you rmdir the downloadDestination first?

Reply:
Right now I remove files in downloadFile, I think I agree that I like this approach more though.

Reply:
Updated accordingly.